src/Domain/Services/CreateAccount.php

```diff
@@ -0,0 +1,233 @@
+					function( $customer_id, $new_customer_data = array(), $password_generated = false ) use ( $wc_emails_instance ) {
+						// If this is a block-based signup, send a new email
+						// with password reset link (no password in email).
+						if ( isset( $new_customer_data['is_checkout_block_customer_signup'] ) ) {
+							$this->customer_new_account( $customer_id, $new_customer_data );
+							return;
+						}
+
+						// Otherwise, trigger the existing legacy email (with new password inline).
+						$wc_emails_instance->customer_new_account( $customer_id, $new_customer_data, $password_generated );
+					},
+					10,
+					3
+				);
+			}
+		);
+	}
+
+	/**
+	 * Trigger new account email.
+	 * This is intended as a replacement to WC_Emails::customer_new_account(),
+	 * with a set password link instead of emailing the new password in email
+	 * content.
+	 *
+	 * @param int   $customer_id       The ID of the new customer account.
+	 * @param array $new_customer_data Assoc array of data for the new account.
+	 */
+	public function customer_new_account( $customer_id = 0, array $new_customer_data = array() ) {
+		if ( ! $customer_id ) {
+			return;
+		}
+
+		if ( ! $this->new_account_email ) {
+			return;
+		}
+
+		$this->new_account_email->trigger( $customer_id, $new_customer_data );
+	}
+
+	/**
+	 * Create a user account for specified order and request (if necessary).
+	 * If a new account is created:
+	 * - The order is associated with the account.
+	 * - The user is logged in.
+	 *
+	 * @param \WC_Order        $order   The order currently being processed.
+	 * @param \WP_REST_Request $request The current request object being handled.
+	 *
+	 * @throws Exception On error.
+	 */
+	public function from_order_request( \WC_Order $order, \WP_REST_Request $request ) {
+		if ( ! $this->should_create_customer_account( $request ) ) {
+			return;
+		}
+
+		$customer_id = $this->create_customer_account(
+			$order->get_billing_email(),
+			$order->get_billing_first_name(),
+			$order->get_billing_last_name()
+		);
+
+		// Log the customer in and associate with the order.
+		wc_set_customer_auth_cookie( $customer_id );
+		$order->set_customer_id( get_current_user_id() );
+	}
+
+	/**
+	 * Check request options and store (shop) config to determine if a user account
+	 * should be created as part of order processing.
+	 *
+	 * @param \WP_REST_Request $request The current request object being handled.
+	 *
+	 * @return boolean True if a new user account should be created.
+	 */
+	protected function should_create_customer_account( $request ) {
+		if ( is_user_logged_in() ) {
+			return false;
+		}
+
+		$checkout_requires_account = false === filter_var( get_option( 'woocommerce_enable_guest_checkout' ), FILTER_VALIDATE_BOOLEAN );
+		if ( $checkout_requires_account ) {
+			return true;
+		}
+
+		$user_requested_account = ! empty( $request['should_create_account'] ) && true === filter_var( $request['should_create_account'], FILTER_VALIDATE_BOOLEAN );
```

Comment:
You could probably simplify by doing something like this:

```php
protected function should_create_customer_account( $request ) {
	if ( is_user_logged_in() || empty( $request['should_create_account'] ) ) {
		return false;
	}

	if ( 
		false === filter_var( get_option( 'woocommerce_enable_guest_checkout' ), FILTER_VALIDATE_BOOLEAN ) ||
		true === filter_var( $request['should_create_account'], FILTER_VALIDATE_BOOLEAN )
	) {
		return true;
	}

	return false;
}
```

Reply:
There's some complexity to this - depending on store options, we may make an account irrespective of what user requests. I'll revamp so it's clear and simplify as much as possible 😄

Reply:
Updated and commented this here: e95bf7f9b86ab8033ee2a8174e8fb3175ea09087